apollo-router/src/plugins/telemetry/dynamic_attribute.rs

```diff
@@ -0,0 +1,159 @@
+            | LogAttributes::Subgraph(attributes)
+            | LogAttributes::Supergraph(attributes) => attributes,
+        }
+    }
+
+    fn insert(&mut self, span_name: &str, key: String, value: String) {
+        match span_name {
+            ROUTER_SPAN_NAME => {
+                if let Self::Router(attributes) = self {
+                    attributes.insert(key, value);
+                }
+            }
+            SUBGRAPH_SPAN_NAME => {
+                if let Self::Subgraph(attributes) = self {
+                    attributes.insert(key, value);
+                }
+            }
+            SUPERGRAPH_SPAN_NAME => {
+                if let Self::Supergraph(attributes) = self {
+                    attributes.insert(key, value);
+                }
+            }
+            _ => {
+                eprintln!("cannot add custom attributes to this span '{span_name}', it's only available on router/supergraph/subgraph spans");
+            }
+        }
+    }
+    fn extend(&mut self, span_name: &str, val: impl IntoIterator<Item = (String, String)>) {
+        match span_name {
+            ROUTER_SPAN_NAME => {
+                if let Self::Router(attributes) = self {
+                    attributes.extend(val);
+                }
+            }
+            SUBGRAPH_SPAN_NAME => {
+                if let Self::Subgraph(attributes) = self {
+                    attributes.extend(val);
+                }
+            }
+            SUPERGRAPH_SPAN_NAME => {
+                if let Self::Supergraph(attributes) = self {
+                    attributes.extend(val);
+                }
+            }
+            _ => {
+                eprintln!("cannot add custom attributes to this span '{span_name}', it's only available on router/supergraph/subgraph spans");
+            }
+        }
+    }
+}
+
+pub(crate) struct DynAttributeLayer;
+
+impl<S> Layer<S> for DynAttributeLayer
+where
+    S: tracing_core::Subscriber + for<'lookup> tracing_subscriber::registry::LookupSpan<'lookup>,
+{
+    fn on_new_span(
+        &self,
+        _attrs: &tracing_core::span::Attributes<'_>,
+        id: &tracing_core::span::Id,
+        ctx: Context<'_, S>,
+    ) {
+        let span = ctx.span(id).expect("Span not found, this is a bug");
+        let custom_attributes = match span.name() {
+            ROUTER_SPAN_NAME => LogAttributes::Router(HashMap::new()),
+            SUBGRAPH_SPAN_NAME => LogAttributes::Subgraph(HashMap::new()),
+            SUPERGRAPH_SPAN_NAME => LogAttributes::Supergraph(HashMap::new()),
+            _ => {
+                return;
+            }
+        };
+        let mut extensions = span.extensions_mut();
+        if extensions.get_mut::<LogAttributes>().is_none() {
+            extensions.insert(custom_attributes);
+        }
+    }
+}
+
+impl DynAttributeLayer {
+    pub(crate) fn new() -> Self {
+        Self {}
+    }
+}
+
+pub(crate) trait DynAttribute {
+    fn set_dyn_attribute(&self, key: String, value: String);
```

Comment:
nit: it's likely that keys will stay the same for the entire time config and schema will be used, so you could avoid allocations by using `Arc<String>`

Reply:
Yes related to the same comment [here](https://github.com/apollographql/router/pull/4102#discussion_r1373298243)